data/transactions/logic/frames.go

```diff
@@ -0,0 +1,101 @@
+// Copyright (C) 2019-2022 Algorand, Inc.
+// This file is part of go-algorand
+//
+// go-algorand is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// go-algorand is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with go-algorand.  If not, see <https://www.gnu.org/licenses/>.
+
+package logic
+
+import (
+	"errors"
+	"fmt"
+)
+
+func opProto(cx *EvalContext) error {
+	if !cx.fromCallsub {
+		return fmt.Errorf("proto executed in normal flow")
+	}
+	cx.fromCallsub = false
+	nargs := int(cx.program[cx.pc+1])
+	if nargs > len(cx.stack) {
+		return fmt.Errorf("callsub to proto that requires %d args with stack height %d", nargs, len(cx.stack))
+	}
+	top := len(cx.callstack) - 1
+	cx.callstack[top].clear = true
+	cx.callstack[top].args = nargs
+	cx.callstack[top].returns = int(cx.program[cx.pc+2])
+	return nil
+}
+
+func opFrameDig(cx *EvalContext) error {
+	i := int8(cx.program[cx.pc+1])
+
+	top := len(cx.callstack) - 1
+	if top < 0 {
+		return errors.New("frame_dig with empty callstack")
+	}
+
+	frame := cx.callstack[top]
+	if i < 0 && -int(i) > frame.args {
```

Comment:
the first clause `i < 0` of the conjunction appears redundant unless there is an intent to short circuit for efficiency's sake (but I doubt there would be speedup overall)

Reply:
That is true. My intention was to make it very clear: "This is the negative `i` case!"

This case is also the one I should think more about. I should probably _not_ perform this check if `proto` wasn't used in the subroutine, at least according to my reasoning above.

Reply:
I changed it to this.
```
	// If proto was used, don't allow `frame_dig` to go below specified args
	if frame.clear && -int(i) > frame.args {
		return fmt.Errorf("frame_dig %d in sub with %d args", i, frame.args)
	}

```
`clear` is used to say that the frame has been marked by `proto`, so we know the number of args.